src/sftp.rs

```diff
@@ -0,0 +1,400 @@
+use super::{Error, Session};
+use std::io::{self, prelude::*};
+use std::process::Stdio;
+
+// TODO: it would _probably_ be better to actually use sftp here, since I'm pretty sure it has some
+// kind of protocol optimized for binary data. but for now, this is fine.
+
+/// A file-oriented channel to a remote host.
+///
+/// You likely want [`write_to`] and [`read_from`].
+#[derive(Debug, Clone)]
+pub struct Sftp<'s> {
+    session: &'s Session,
+}
+
+/// A file access mode.
+#[derive(Debug, Clone, Copy)]
+pub enum Mode {
+    /// Read-only access.
+    Read,
+    /// Write-only access.
+    Write,
+    /// Write-only access in append mode.
+    Append,
+}
+
+impl Mode {
+    fn is_write(self) -> bool {
+        match self {
+            Mode::Append | Mode::Write => true,
+            Mode::Read => false,
+        }
+    }
+}
+
+/// A handle to a remote file.
+///
+/// If you opened this file for reading (with [`Sftp::read_from`]), you can read from it with
+/// [`std::io::Read`]. If you opened it for writing (with [`Sftp::write_to`]), you can write to it
+/// with [`std::io::Write`].
+///
+/// Note that because we are operating against a remote host, errors may take a while to propagate.
+/// For that reason, you may find that writes to, and reads from, files that do not exist, may
+/// still succeed without error. You should **make sure to call [`close`]** to observe any errors
+/// that may have occurred when operating on the remote file.
+#[derive(Debug)]
+pub struct RemoteFile<'s> {
+    cat: super::RemoteChild<'s>,
+    mode: Mode,
+}
+
+// TODO: something like std::fs::OpenOptions
+
+impl<'s> Sftp<'s> {
+    pub(crate) fn new(session: &'s Session) -> Self {
+        Self { session }
+    }
+
+    // TODO: remove
+    // TODO: exists
+
+    /// Check that the given file can be opened in the given mode.
+    ///
+    /// Note that this function is potentially racy. The permissions on the server could change
+    /// between when you check and when you start a subsequent operation, causing it to fail. It
+    /// can also produce false positives, such as if you are checking whether you can write to a
+    /// file on a read-only file system that you still have write permissions on.
+    pub fn can(&mut self, mode: Mode, path: impl AsRef<std::path::Path>) -> Result<(), Error> {
```

Comment:
Why `pub`?

Reply:
In case someone wants to check permissions separately from doing an operation?

Reply:
yeah, makes sense - see my comment above about maybe giving some guidance about when to use each